### PR TITLE
Sequence cache support

### LINKED ISSUE #11261: Support cache when creating a sequence
Hi,
    May I know are there any plan for ef core to support cache size for sequence, something like this:
modelBuilder.HasSequence<int>("OrderNumbers", schema: "shared")
            .StartsAt(1000)
            .IncrementsBy(5)
            **_.CacheSize(10)_**;

Thanks,
Bruce'

### PR DESCRIPTION
**Pull request by bikbov**
- [x] I've read the guidelines for [contributing](CONTRIBUTING.md) and seen the [walkthrough](https://youtu.be/9OMxy1wal1s?t=1869)
- [x] I've posted a comment on an issue with a detailed description of how I am planning to contribute and got approval from a member of the team
- [x] The code builds and tests pass locally (also verified by our automated build checks)
- [x] Commit messages follow this format:

        Summary of the changes
        - Sequence cache support added

        Fixes #11261 

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Code follows the same patterns and style as existing code in this repo

Fixes #11261

### CODE REVIEW

File: src/EFCore.Relational/Migrations/MigrationBuilder.cs [hunk 1]
```diff
@@ -518,10 +518,14 @@ public MigrationBuilder(string? activeProvider)
...
+    /// <param name="cached">Indicates whether the sequence use preallocated values.</param>
+    /// <param name="cacheSize">The amount of preallocated values.</param>
     /// <param name="oldIncrementBy">The previous amount to increment by when generating the next value in the sequence, defaulting to 1.</param>
     /// <param name="oldMinValue">The previous minimum value of the sequence, or <see langword="null" /> if not specified.</param>
     /// <param name="oldMaxValue">The previous maximum value of the sequence, or <see langword="null" /> if not specified.</param>
     /// <param name="oldCyclic">Indicates whether or not the sequence would previously re-start when the maximum value is reached.</param>
+    /// <param name="oldCached">Indicates whether the sequence previously used preallocated values</param>
+    /// <param name="oldCacheSize">The previous amount of preallocated values.</param>
     /// <returns>A builder to allow annotations to be added to the operation.</returns>
     public virtual AlterOperationBuilder<AlterSequenceOperation> AlterSequence(
```
Comment (ajcvickers) [inline src/EFCore.Relational/Migrations/MigrationBuilder.cs:530]: Adding parameters like this is a breaking change. @bricelam What do we normally do in these cases?

File: test/EFCore.SqlServer.FunctionalTests/Migrations/MigrationsSqlServerTest.cs [hunk 2]
```diff
@@ -2649,7 +2682,67 @@ public override async Task Alter_sequence_increment_by()
...
+""");
+    }
+
+    public override async Task Alter_sequence_cache_to_default_cache()
+    {
+        await base.Alter_sequence_cache_to_default_cache();
+
+        AssertSql(
+"""
+ALTER SEQUENCE [Eta] INCREMENT BY 1 NO MINVALUE NO MAXVALUE NO CYCLE CACHE;
```
Comment (ajcvickers) [inline test/EFCore.SqlServer.FunctionalTests/Migrations/MigrationsSqlServerTest.cs:2725]: @bricelam @AndriySvyryd Should we only include "CACHE" explicitly here if the user has explicitly specified? If they don't specify anything, then should we leave it off and let SQL Server do its default?

File: src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs [hunk 3]
```diff
@@ -804,6 +804,19 @@ valueGenerated switch
             root = root?.Chain(isCyclic) ?? isCyclic;
         }
 
+        if (sequence.IsCached != Sequence.DefaultIsCached)
```
Comment (ajcvickers) [inline src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:807]: @bricelam @AndriySvyryd Should we only scaffold something here if the user has explicitly set some caching options?

**Comment by ajcvickers** [comment]
@bikbov Mostly looks good. I will follow up with the team on the open questions.

File: test/EFCore.SqlServer.FunctionalTests/Migrations/MigrationsSqlServerTest.cs (hunk 2 above)
Comment (roji) [reply test/EFCore.SqlServer.FunctionalTests/Migrations/MigrationsSqlServerTest.cs:2725]: That would be my vote, yes.

**Comment by roji** [comment]
Note some previous discussion in https://github.com/dotnet/efcore/issues/11261#issuecomment-1280804583 and above

File: src/EFCore.Relational/Migrations/MigrationBuilder.cs (hunk 1 above)
Comment (bricelam) [reply src/EFCore.Relational/Migrations/MigrationBuilder.cs:530]: Do cached sequences belong at the Relational level? I thought they were SQL Server-specific. If so, these would come through as annotations on the operations, not new API.

Comment (roji) [reply src/EFCore.Relational/Migrations/MigrationBuilder.cs:530]: I think all the databases which support sequences support cache as well ([PostgreSQL](https://www.postgresql.org/docs/current/sql-createsequence.html), [SQL Server](https://learn.microsoft.com/en-us/sql/t-sql/statements/create-sequence-transact-sql?view=sql-server-ver16), [Oracle](https://docs.oracle.com/cd/B12037_01/server.101/b10759/statements_6014.htm)). Nice to see some consistency every now and then :)

File: src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:807]: Answer from team: yes.

File: test/EFCore.SqlServer.FunctionalTests/Migrations/MigrationsSqlServerTest.cs (hunk 2 above)
Comment (ajcvickers) [reply test/EFCore.SqlServer.FunctionalTests/Migrations/MigrationsSqlServerTest.cs:2725]: Team agreement on this.

File: src/EFCore.Relational/Migrations/MigrationBuilder.cs (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore.Relational/Migrations/MigrationBuilder.cs:530]: @bricelam Any comment on the addition of parameters here?

Comment (ajcvickers) [reply src/EFCore.Relational/Migrations/MigrationBuilder.cs:530]: Note from triage: should be okay to add them add the end, making sure they have default values.

**Comment by bikbov** [comment]
@ajcvickers  Thanks for review. I will update pull request

**Comment by Saibamen** [comment]
@bikbov: ping

**Review by ajcvickers — APPROVED** [review]

**Comment by ajcvickers** [comment]
@bikbov Really sorry for taking so long on this PR. I completed the final two items [and merged](https://github.com/dotnet/efcore/commit/73904c0fe00185895c9a6574a8f043ed5e162f34) after squashing a rebasing. This PR will show up as closed, but the commit in the main branch is still credited to you. Again, sorry for taking so long, and thanks for you hard work on this. It ended up being a high-quality contribution.

**Comment by bikbov** [comment]
@ajcvickers Thank you

**Pull request closed without merging**